src/Elastic.Transport/Components/Serialization/SystemTextJsonSerializer.cs

```diff
@@ -0,0 +1,168 @@
+
+using System;
+using System.IO;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Elastic.Transport;
+
+/// <summary>
+/// An abstract implementation of a transport <see cref="Serializer"/> which serializes using the Microsoft
+/// <c>System.Text.Json</c> library.
+/// </summary>
+public abstract class SystemTextJsonSerializer :
+	Serializer
+{
+	private readonly SemaphoreSlim _semaphore = new(1);
+
+	private bool _initialized;
+	private JsonSerializerOptions? _options;
+	private JsonSerializerOptions? _indentedOptions;
+
+	#region Serializer
+
+	/// <inheritdoc />
+	public override T Deserialize<T>(Stream stream)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out T deserialize))
+			return deserialize;
+
+		return JsonSerializer.Deserialize<T>(stream, _options);
+	}
+
+	/// <inheritdoc />
+	public override object? Deserialize(Type type, Stream stream)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out object deserialize))
+			return deserialize;
+
+		return JsonSerializer.Deserialize(stream, type, _options);
+	}
+
+	/// <inheritdoc />
+	public override ValueTask<T> DeserializeAsync<T>(Stream stream, CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out T deserialize))
+			return new ValueTask<T>(deserialize);
+
+		return JsonSerializer.DeserializeAsync<T>(stream, _options, cancellationToken);
+	}
+
+	/// <inheritdoc />
+	public override ValueTask<object?> DeserializeAsync(Type type, Stream stream, CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out object deserialize))
+			return new ValueTask<object?>(deserialize);
+
+		return JsonSerializer.DeserializeAsync(stream, type, _options, cancellationToken);
+	}
+
+	/// <inheritdoc />
+	public override void Serialize<T>(T data, Stream writableStream,
+		SerializationFormatting formatting = SerializationFormatting.None)
+	{
+		Initialize();
+
+		JsonSerializer.Serialize(writableStream, data, GetJsonSerializerOptions(formatting));
+	}
+
+	/// <inheritdoc />
+	public override Task SerializeAsync<T>(T data, Stream stream,
+		SerializationFormatting formatting = SerializationFormatting.None,
+		CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		return JsonSerializer.SerializeAsync(stream, data, GetJsonSerializerOptions(formatting), cancellationToken);
+	}
+
+	#endregion Serializer
+
+	/// <summary>
+	/// A factory method that can create an instance of <see cref="JsonSerializerOptions"/> that will
+	/// be used when serializing.
+	/// </summary>
+	/// <returns></returns>
+	protected abstract JsonSerializerOptions? CreateJsonSerializerOptions();
+
+	/// <summary>
+	/// A callback function that is invoked after the <see cref="JsonSerializerOptions"/> have been created and the
+	/// serializer got fully initialized.
+	/// </summary>
+	protected virtual void Initialized()
```

Comment:
The `DefaultSourceSerializer` in `Elastic.Clients.Elasticsearch` registers the `JsonSerializerOptions` in a map for internal use. This callback method will be used for this purpose.